REGRESSION(282068@main): Scrollbars no longer match YouTube's dark theme

nmoucht opened the pull request:
#### 7bfadd7e9dc6f12de05d72357d6dad3b8ed18b2d
<pre>
REGRESSION(282068@main): Scrollbars no longer match YouTube&apos;s dark theme
<a href="https://bugs.webkit.org/show_bug.cgi?id=279942">https://bugs.webkit.org/show_bug.cgi?id=279942</a>
&lt;<a href="https://rdar.apple.com/problem/135886305">rdar://problem/135886305</a>&gt;

Reviewed by Tim Nguyen.

Add quirk to youtube to set color-scheme to dark if dark attribute is present on html element. This is due
to scrollbar-width being turned on, which overrides the scrollbar pseudo elements being used by youtube,
which exposes the native scrollbars not being set to dark mode.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsYouTubeDarkModeQuirk const):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):

Canonical link: <a href="https://commits.webkit.org/284010@main">https://commits.webkit.org/284010@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b3697b9c55d37b0cc360653356d924569d1f69c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67951 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47343 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20610 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72115 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19096 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70068 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18912 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12737 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71018 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43361 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58722 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34866 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40028 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16136 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17453 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62001 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16477 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73706 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11922 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15770 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61782 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11960 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58795 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61794 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9683 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3297 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10364 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43143 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44217 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45412 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43959 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -672,6 +672,19 @@ bool Quirks::needsYouTubeOverflowScrollQuirk() const
 }
 
+// FIXME: Remove after the site is fixed, <need>
+// youtube.com rdar://135886305
```
```suggestion
// youtube.com rdar://135886305
```

You could drop the FIXME altogether, the FIXME really applies to all the quirks in this file.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -984,6 +985,16 @@ void Adjuster::adjustForSiteSpecificQuirks(RenderStyle& style) const
+        // Sets color-scheme to dark if dark attribute is applied to root element.
+        if (RefPtr div = dynamicDowncast<HTMLHtmlElement>(m_element)) {
+            static MainThreadNeverDestroyed<const AtomString> darkMode("dark"_s);
+            if (div->hasAttribute(darkMode))
```
```suggestion
        if (m_element && m_element == m_document->documentElement()) {
            static MainThreadNeverDestroyed<const AtomString> darkMode("dark"_s);
            if (m_element->hasAttribute(darkMode))
```

This is probably slightly more performant.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -672,6 +672,19 @@ bool Quirks::needsYouTubeOverflowScrollQuirk() const
+        return false;
+
+    if (!m_needsYouTubeDarkModeQuirk)
+        m_needsYouTubeDarkModeQuirk = m_document->url().host() == "www.youtube.com"_s;
```
What about tv.youtube.com. studio.youtube.com, www.youtubekids.com etc?

nmoucht replied:
I don't think we have evidence that this issue exists on these.

nmoucht replied:
After some looking this issue also exists on studio.youtube.com. Made this for all YouTube.com subdomains.

nt1m commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -41,6 +41,7 @@
 #include "HTMLBodyElement.h"
 #include "HTMLDialogElement.h"
 #include "HTMLDivElement.h"
+#include "HTMLHtmlElement.h"
```
This import is no longer used

nmoucht replied:
Fixed.

nmoucht committed:
REGRESSION(282068@main): Scrollbars no longer match YouTube's dark theme
https://bugs.webkit.org/show_bug.cgi?id=279942
<rdar://problem/135886305>

Reviewed by Tim Nguyen.

Add quirk to youtube to set color-scheme to dark if dark attribute is present on html element. This is due
to scrollbar-width being turned on, which overrides the scrollbar pseudo elements being used by youtube,
which exposes the native scrollbars not being set to dark mode.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsYouTubeDarkModeQuirk const):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):

Canonical link: https://commits.webkit.org/284010@main
